### PR TITLE
fix: en leaks 'error-new-signup-email', 'error-new-signup-email-exists'

### PR DESCRIPTION
**Pull request by sherizada-okta**
## Description:
Fix english leaks for enroll-profile sign up errors.

1. With a known param, "Email", that does get translated:
Screenshot 2021-05-05 at 2 42 05 PM
Screenshot 2021-05-07 at 12 02 14 PM
2. With an unknown param, "Custom Property", that does not get translated:
Screenshot 2021-05-05 at 2 51 53 PM
Screenshot 2021-05-07 at 12 03 26 PM

## PR Checklist

- [x] Have you verified the basic functionality for this change?

- [x] Added unit tests?

- [ ] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Issue:

- [OKTA-389420](https://oktainc.atlassian.net/browse/OKTA-389420)

### CODE REVIEW

File: src/v2/ion/i18nTransformer.js [hunk 1]
```diff
@@ -150,6 +150,50 @@ const I18N_PARAMS_MAPPING = {
...
+ *
+ * Example:
+ * input =
+ *  "i18n": {
+      "key": "registration.error.doesNotMatchPattern",
+      "params": [
+        "Email"
+      ]
+    }
+ * output = registration.error.invalidLoginEmail.Email
```
Comment (nikhilvenkatraman-okta) [inline src/v2/ion/i18nTransformer.js:180]: nit: can we correct this to `registration.error.doesNotMatchPattern.Email` to match the input example ?

File: src/v2/ion/i18nTransformer.js [hunk 2]
```diff
@@ -288,13 +332,22 @@ const isWebAuthnAPIError = ( i18nKey ) => i18nKey.startsWith(WEBAUTHN_API_GENERI
...
-    const i18nKey = I18N_OVERRIDE_MAPPINGS[message.i18n?.key] ?? message.i18n.key;
+    let i18nKey = message.i18n.key;
+    let i18nParms = message.i18n.params || [];
+
+    if (I18N_OVERRIDE_MAPPINGS[message.i18n?.key]) {
+      i18nKey = I18N_OVERRIDE_MAPPINGS[message.i18n?.key];
+    } else if (I18N_OVERRIDE_WITH_PARAMS_MAP[i18nKey]) {
+      const param = message.i18n.params?.[0];
+      i18nKey = getI8nKeyUsingParams(i18nKey, param);
+      i18nParms = [];
```
Comment (nikhilvenkatraman-okta) [inline src/v2/ion/i18nTransformer.js:343]: nit: `i18nParams`


Can you share an example of what the response look like for `registration.error.invalidLoginEmail.custom` ? I see that the value for this key is `'{0}' must be in the form of an email address` but we set `i18nParms` to []. How does the value for `{0}` get applied in this case?

**Review by nikhilvenkatraman-okta — COMMENTED** [review]
Can we add some tests for the new getI8nKeyUsingParams method in i18nTransformer_spec ?

File: src/v2/ion/i18nTransformer.js (hunk 2 above)
Comment (sherizada-okta) [reply src/v2/ion/i18nTransformer.js]: @nikhilvenkatraman-okta good catch, I pushed the commit after your reviewed https://github.com/okta/okta-signin-widget/pull/1912/commits/9968de5cd2efe20d0b60c9b0f4265d5ca9f09445#diff-c64539a9635d1e00f39eaf1a3c3e8b634aed28ae450ad9a1740fefd6076b06dbR346

**Comment by sherizada-okta** [comment]
> Can we add some tests for the new getI8nKeyUsingParams method in i18nTransformer_spec ?

Yes if this approach sounds good?

File: src/v2/ion/i18nTransformer.js (hunk 2 above)
Comment (sherizada-okta) [reply src/v2/ion/i18nTransformer.js]: Documenting the custom case: https://github.com/okta/okta-signin-widget/pull/1912/commits/3798d028dc6397d9474761b51a95377a36abfa28

Also see PR description for screenshots

File: src/v2/ion/i18nTransformer.js [hunk 3]
```diff
@@ -288,13 +347,22 @@ const isWebAuthnAPIError = ( i18nKey ) => i18nKey.startsWith(WEBAUTHN_API_GENERI
...
-    const i18nKey = I18N_OVERRIDE_MAPPINGS[message.i18n?.key] ?? message.i18n.key;
+    let i18nKey = message.i18n.key;
+    let i18nParams = message.i18n.params || [];
+
+    if (I18N_OVERRIDE_MAPPINGS[message.i18n?.key]) {
+      i18nKey = I18N_OVERRIDE_MAPPINGS[message.i18n?.key];
+    } else if (I18N_OVERRIDE_WITH_PARAMS_MAP[i18nKey]) {
+      const param = message.i18n.params?.[0];
+      i18nKey = getI8nKeyUsingParams(i18nKey, param);
+      i18nParams = i18nKey.endsWith('custom') ? [param] : [];
```
Comment (sherizada-okta) [inline src/v2/ion/i18nTransformer.js:358]: need to test this gets polyfilled in ie

File: packages/@okta/i18n/src/properties/login.properties [hunk 4]
```diff
@@ -651,6 +651,12 @@ registration.default.callbackhook.error=We could not process your registration a
 registration.error.userName.invalidEmail=Invalid email address
 registration.error.password.passwordRequirementsNotMet=Password requirements were not met
 registration.error.userName.notUniqueWithinOrg=An account with that email already exists
+registration.error.notUniqueWithinOrg.Email = A user with this 'Email' already exists
```
Comment (nikhilvenkatraman-okta) [inline packages/@okta/i18n/src/properties/login.properties:654]: Are Email and custom attributes the only 2 types of known errors we can run into ?

File: src/v2/ion/i18nTransformer.js (hunk 3 above)
Comment (nikhilvenkatraman-okta) [reply src/v2/ion/i18nTransformer.js:358]: `endsWith` does not seem to have any [IE support ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/endsWith#browser_compatibility)

Comment (pradeepdewda-okta) [reply src/v2/ion/i18nTransformer.js:358]: Do we really have customers who uses IE ? It's 2021 :D

Comment (sherizada-okta) [reply src/v2/ion/i18nTransformer.js:358]: It's polyfilled in IE 
Screenshot 2021-05-07 at 11 59 29 AM

File: packages/@okta/i18n/src/properties/login.properties (hunk 4 above)
Comment (sherizada-okta) [reply packages/@okta/i18n/src/properties/login.properties:654]: Yes, when testing out Profile enrollment form fields, I found that 'Email' is the only pre-filled label.

**Comment by sherizada-okta** [comment]
Ready for another review. Added unit tests.

File: src/v2/ion/i18nTransformer.js (hunk 3 above)
Comment (nikhilvenkatraman-okta) [reply src/v2/ion/i18nTransformer.js:370]: @pradeepdewda-okta  we have customers who load the widget inside embedded browsers not just IE 😄

**Review by nikhilvenkatraman-okta — APPROVED** [review]

**Pull request closed without merging**